File: 20. Valid Parentheses/isValid.cpp
#include <stack>
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        //edge case, odd length of input string is automatically invalid
        if(s.length() % 2 == 1) {
            return false;
        }
        
        stack<char> recordStack;
        for(int i = 0; i < s.length(); i++) {
            if(s[i] == '(' || s[i] == '[' || s[i] == '{') {
                recordStack.push(s[i]);
            } else {
                switch (s[i]) {
                    case ')': {
                        if(recordStack.pop() != '(') {
                            return false;
                        }
                        break;
                    }
                    case ']': {
                        if(recordStack.pop() != '[') {
                            return false;
                        }
                        break;
                    }
                    case '}': {
                        if(recordStack.pop() != '{') {
                            return false;
                        }
                    }
                }
            }
        }
        if(recordStack.size() > 0) {
            return false;
        }
        return true;
    }
};

int main() {
    Solution sol;

    string s1 = "()";
    bool output1 = sol.isValid(s1);
    cout << "Test 1: Expecting true (1) got " << output1 << "\n";


    string s2 = "()[]{}";
    bool output2 = sol.isValid(s2);
    cout << "Test 2: Expecting true (1) got " << output2 << "\n";

    string s3 = "(]";
    bool output3 = sol.isValid(s3);
    cout << "Test 3: Expecting false (0) got " << output3 << "\n";

    string s4 = "(((([[[[{{{{}}}}]]]]))))";
    bool output4 = sol.isValid(s4);
    cout << "Test 4: Expecting true (1) got " << output4 << "\n";

    string s5 = "(()()(){}{}{}{}{{{{{}}]][}";
    bool output5 = sol.isValid(s5);
    cout << "Test 5: Expecting false (0) got " << output5 << "\n";

    string s6 = "(";
    bool output6 = sol.isValid(s6);
    cout << "Test 6 (string input length 1): Expecting false (0) got " << output6 << "\n";

    string s7 = "((";
    bool output7 = sol.isValid(s7);
    cout << "Test 7: Expecting false (0) got " << output7 << "\n";

    string s8 = ")";
    bool output8 = sol.isValid(s8);
    cout << "Test 8: Expecting false (0) got " << output8 << "\n";

    string s9 = "){";
    bool output9 = sol.isValid(s9);
    cout << "Test 9: Expecting false (0) got " << output9 << "\n";

}

/* Intuition: Given an input string that comprises only the characters '(', ')', '{', '}', '[' and ']' we are asked to determine if the input string is valid based on if the parentheses (bracket, curly or normal) are the appropriate type and in order. We need a way to store the opening bracket and the order each opening bracket appeared in. When we start seeing the closing brackets again, look at the most recent think added and see if it matches the expected. A stack structure would seem to work best for this purpose. 

Approach:  Declare an array to store the record of opening brackets. Begin iterating through the input string, and check if it is an opening bracket, if it is, add that character into the next part of the recordArray. If instead a closing bracket is encountered, we compare the last added bracket in the recordArray and compare it to the closing bracket (i.e. if we have ']' then '[' should be on record) If at any time that comparison fails, return false. After iteration through the input string, do a final check of the tracker index, it should be at 0. If it is not, return false because we didn't fully close out the opening brackets. Otherwise return true.

Complexity: 
-Time: We loop through the input array operations once so the time complexity is $$O(n)$$
-Space: The record array could be as large as the input array at the most, so space complexity is $$O(n)$$.  */

File: 20. Valid Parentheses/isValid.js
/**
 * @param {string} s
 * @return {boolean}
 */
/*
var isValid = function(s) {
    //edge case: if input string is odd length it cannot be valid
    if(s.length % 2 === 1) {
        return false;
    }

    const recordArr = [];
    let poppedBracket;
    for(bracket of s) {
        // if it's an opening bracket, just push onto the record array
        if(bracket === '(' || bracket === '{' || bracket === '[') {
            recordArr.push(bracket);
        } else {
            // we know that bracket is a closing one, so get the most recent opening bracket from the record array
            poppedBracket = recordArr.pop(); 
            switch(bracket) {
                case ')': {
                    if(poppedBracket != '(') {
                        return false;
                    } 
                    break;
                }
                case '}': {
                    if(poppedBracket != '{') {
                        return false;
                    }
                    break;
                }
                case ']': {
                    if(poppedBracket != '[') {
                        return false;
                    }
                    break;
                }
            }
        }
    }
    return (recordArr.length === 0);
};
*/
/*Internet solution: Is more elegant because it uses an object as a hash table to determine what closing bracket is associated with each opening bracket */
var isValid = function(s) {
    //edge case: if input string is odd length it cannot be valid
    if(s.length % 2 === 1) {
        return false;
    }
    const pairs = {
        "(": ")",
        "[": "]",
        "{": "}"
    }
    const recordArr = [];
    for(let bracket of s) {
        // if it's an opening bracket, just push onto the record array
        if(bracket === '(' || bracket === '{' || bracket === '[') {
            recordArr.push(bracket);
        } else if(pairs[recordArr.pop()] !== bracket){
            // we know that bracket is a closing one, so get the most recent opening bracket from the record array
            return false;
        }
    }
    return (recordArr.length === 0);
};


/* Intuition: Given an input string that comprises only the characters '(', ')', '{', '}', '[' and ']' we are asked to determine if the input string is valid based on if the parentheses (bracket, curly or normal) are the appropriate type and in order. We need a way to store the opening bracket and the order each opening bracket appeared in. When we start seeing the closing brackets again, look at the most recent think added and see if it matches the expected. A stack structure would seem to work best for this purpose. 

Approach: Javascript arrays with pop and push function as a stack so we will declare an array to store the record of opening brackets. Begin iterating through the input string, and check if it is an opening bracket, if it is, push that character onto the record array. If instead a closing bracket is encountered, pop the last element from the array and compare it to the closing bracket (i.e. if we have ']' then '[' should have been popped, etc.) If at any time that comparison fails, return false. After iteration through the input string, do a final check of the length of the record array. If it has any elements remaining, return false because we didn't fully close out the opening brackets. Otherwise return true.

Complexity: 
-Time: We loop through the input array operations once so the time complexity is $$O(n)$$
-Space: The record array could be as large as the input array at the most, so space complexity is $$O(n)$$.  */


const s1 = "()";
const output1 = isValid(s1);
console.log("Test 1: Expecting true got " + output1);


const s2 = "()[]{}";
const output2 = isValid(s2);
console.log("Test 2: Expecting true got " + output2);

const s3 = "(]";
const output3 = isValid(s3);
console.log("Test 3: Expecting false got " + output3);

const s4 = "(((([[[[{{{{}}}}]]]]))))";
const output4 = isValid(s4);
console.log("Test 4: Expecting true got " + output4);

const s5 = "(()()(){}{}{}{}{{{{{}}]][}";
const output5 = isValid(s5);
console.log("Test 5: Expecting false got " + output5);

const s6 = "(";
const output6 = isValid(s6);
console.log("Test 6 (string input length 1): Expecting false got " + output6);

const s7 = "((";
const output7 = isValid(s7);
console.log("Test 7: Expecting false got " + output7);

const s8 = ")";
const output8 = isValid(s8);
console.log("Test 8: Expecting false got " + output8);